Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -356,12 +357,41 @@ public struct BSONDocument {
         /// Appends element to underlying BSON bytes, returns the size of the element appended: type + key + value
         @discardableResult internal mutating func append(key: String, value: BSON) -> Int {
             let writer = self.buffer.writerIndex
-            self.buffer.writeInteger(value.bsonValue.bsonType.rawValue, as: UInt8.self)
-            self.buffer.writeCString(key)
+            self.appendElementHeader(key: key, bsonType: value.bsonValue.bsonType)
             value.bsonValue.write(to: &self.buffer)
             return self.buffer.writerIndex - writer
         }
 
+        /// Append the header (key and BSONType) for a given element.
+        @discardableResult internal mutating func appendElementHeader(key: String, bsonType: BSONType) -> Int {
+            let writer = self.buffer.writerIndex
+            self.buffer.writeInteger(bsonType.rawValue, as: UInt8.self)
+            self.buffer.writeCString(key)
+            return self.buffer.writerIndex - writer
+        }
+
+        /// Build a document at the current position in the storage via the provided closure which returns
+        /// how many bytes it wrote.
+        ///
+        /// This may be used to build up a fresh document or a subdocument.
+        internal mutating func buildDocument(_ appendFunc: (inout Self) throws -> Int) rethrows -> Int {
+            var totalBytes = 0
+
+            // write length of document
+            let lengthIndex = self.buffer.writerIndex
```

Comment:
thanks, that is helpful! 

I think what I was imagining with my concern here and the other comment was that someone could do something like
```swift
var doc: BSONDocument = ["a": 1]
doc.storage.buildDocument { storage in
    // add some stuff
}
```
in which case I don't think we'd end up with a valid top level doc, but we aren't doing that anywhere.

Reply:
Oh I see what you mean. Yeah that would be an issue; this method is really only designed to be used during document creation. To prevent something like this at the type level, we could introduce a `BSONDocumentStorageBuilder` or something, but I imagine it'd have some overlap / duplication with the actual `BSONDocumentStorage`.